README.md
## Welcome to <img src="https://raw.githubusercontent.com/ClointFusion/Image_ICONS_GIFs/main/Cloint-LOGO-New.png" height="30"> , Made in India with &#10084;&#65039; 

<br>

<img src="https://raw.githubusercontent.com/ClointFusion/Image_ICONS_GIFs/main/CCEW.PNG">


## Description

Cloint India Pvt. Ltd - Python functions for Robotic Process Automation shortly `RPA`.

# What is ClointFusion?

ClointFusion is an Indian firm based in Vadodara, Gujarat. ClointFusion is a Python-based RPA platform for developing Software BOTs. Using AI, we're working on Common Man's RPA.

#### Check out Project Status

![PyPI](https://img.shields.io/pypi/v/ClointFusion?label=PyPI%20Version) 
![PyPI - License](https://img.shields.io/pypi/l/ClointFusion?label=License) 
![PyPI - Status](https://img.shields.io/pypi/status/ClointFusion?label=Release%20Status) 
![ClointFusion](https://snyk.io/advisor/python/ClointFusion/badge.svg) 
![PyPI - Downloads](https://img.shields.io/pypi/dm/ClointFusion?label=PyPI%20Downloads) 
![Libraries.io SourceRank](https://img.shields.io/librariesio/sourcerank/pypi/ClointFusion) 
![PyPI - Format](https://img.shields.io/pypi/format/ClointFusion?label=PyPI%20Format) 
![GitHub contributors](https://img.shields.io/github/contributors/ClointFusion/ClointFusion?label=Contributors) 
![GitHub last commit](https://img.shields.io/github/last-commit/ClointFusion/ClointFusion?label=Last%20Commit) 

![GitHub Repo stars](https://img.shields.io/github/stars/ClointFusion/ClointFusion?label=Stars&style=social) 
![Twitter URL](https://img.shields.io/twitter/url?style=social&url=https%3A%2F%2Ftwitter.com%2FClointFusion) 
![YouTube Channel Subscribers](https://img.shields.io/youtube/channel/subscribers/UCIygBtp1y_XEnC71znWEW2w?style=social) 
![Twitter Follow](https://img.shields.io/twitter/follow/ClointFusion?style=social)

## Release Notes

- Click here for <a href="https://github.com/ClointFusion/ClointFusion/blob/master/Release_Notes.txt" target="_blank"> Release Notes</a>

---

# Installation

<br>

> ### ClointFusion is now supported on Windows / Ubuntu / macOS* !

## Windows :

* ClointFusion is compatible with both Windows 10 and Windows 11.
* Installing on a Windows PC is a breeze.
* Make certain that Python 3.8 or Python 3.9 is installed.
* Then, from the command prompt, execute the following command.

    ```
    pip install -U ClointFusion
    ```

## Ubuntu :
* Clointfusion requires sudo rights to install on Ubuntu.
* Additional Linux packages must be installed before Clointfusion can be installed.
* Make certain that Python 3.8 or Python 3.9 is installed.
* Then, from the command prompt, execute the following command.

    ```
    sudo apt-get install python3-tk python3-dev
    sudo pip3 install ClointFusion
    ```

<br>

# Importing

<br>

> ### ClointFusion can be accessed using one of two methods.

## Windows :

* ### Terminal : Opens a Python interpreter using the command " import ClointFusion as cf "

    ```
    cf_py
    ```
* ### Code Editor or IDE : Import ClointFusion first, and then run the file in Python.

    ```
    # cf_bot.py

    import ClointFusion as cf

    cf.browser_activate()
    ```
    ```
    python cf_bot.py
    ```


## Ubuntu :

* ### Terminal : Opens a Python interpreter with the command " import ClointFusion as cf " and the required sudo privileges.

    ```
    sudo cf_py
    ```
* ### Code Editor or IDE : Run the file with sudo permissions.

    ```
    # cf_bot.py
    
    import ClointFusion as cf

    cf.browser_activate()
    ```
    ```
    sudo python3 cf_bot.py
    ```

<br>

# Features

<br>

> ## ClointFusion's Voice-Guided, Fully Automated Self-Test.

When you import ClointFusion for the first time, or upgrade to a new version, you'll be prompted with the "ClointFusion Automated Self-Test," which highlights all of ClointFusion's 100+ features in action on your computer while also confirming ClointFusion's compatibility with your PC's settings and configurations.
Once you have successfully completed the self-test, you will receive an email with a self-test report.

Below is the speed up version of self-test.

[ Click here to watch the Self-Test in Action. ](https://user-images.githubusercontent.com/67296473/139620682-d63f6ee6-a3f5-4ca9-9ea9-23216e571e3e.mp4)

<br>

<br>

> ## DOST : Your friend in automation || Build RPA Bots without Code.

` DOST ` is an interactive Blockly based ``no-code`` BOT Builder platform built and optimized for ClointFusion-based BOT building. We feel that automation is important for people other than programmers. Using DOST, even a common man can create a BOT in minutes.

### Advantages of DOST

- Easy to Use.
- Build BOT in minutes.
- No prior Programming knowledge needed.

### Usage of DOST

Open your favorite terminal and type `dost`. That's it!

<img src="https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/VERSION%201.1.0/CLI_DOST_GIF.gif?raw=true" height="400">

**Build BOT with DOST:** [DOST Website](https://dost.clointfusion.com/)

<br>

<br>

> ## BOL : BOL - Your automation voice based assistant

` BOL ` is voice based automation assistant designed to execute BOTs build out of ClointFusion without any human computer interaction.

### Usage of BOL

Open your favorite terminal and type `bol` or `sudo bol` for ubuntu users. Within a moment, a personalized Virtual Assistant will be at your service.

<img src="https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/VERSION%201.1.0/CLI_BOL_GIF.gif?raw=true" height="400">

***Note: `BOL` is currently in development stage. More functionalities are yet to be added.***

<br>

<br>

> ## WORK - The Work Hour Monitor

`WORK` is an intelligent application that detects each and every work you do in your PC and displays a detailed work report.

### Usage of WORK

Open your favorite terminal and type `cf_work`. 
A detailed work report will be displayed.


<img src="https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/VERSION%201.1.0/CLI_CF_WORK.gif?raw=true" height="400">

***Note: All the information that is being collected by `WORK` is stored in a securely maintained database in your system.***

<br>

<br>

> ## WhatsApp Bot - Send bulk WhatsApp messages.

ClointFusion's "WhatsApp Bot" is an automated utility tool that allows you to send many customized messages to your contacts at once.

### Usage of WhatsApp Bot

Open your favorite terminal and type `cf_wm`, and give path of the excel, 
or `cf_wm -e excel_path.xlsx`

[ Click here to watch the WhatsApp Bot in Action. ](https://user-images.githubusercontent.com/67296473/139722199-37036526-2b1c-4120-a12d-bde3df2eb0d7.mp4)

<br>

# ClointFusion in Action

## **Now access more than 100 functions (hit ctrl+space in your IDE)**

***TIP: You can find and inspect all of ClointFusion's functions using only one function i.e., `find()`. Just pass the partial name of the function.***

```
cf.find("sort")

cf.find("gui")

```
* ### 6 gui functions, to take any input from user:

| Function | Accepted Parameters | Description |
| :--------: | :----: | :----------- |
| cf.gui_get_any_input_from_user() | msgForUser="Please enter : ", password=False, multi_line=False, mandatory_field=True | Generic function to accept any input (text / numeric) from user using GUI. Returns the value in string format. |
| cf.gui_get_any_file_from_user() | msgForUser="the file : ", Extension_Without_Dot="*" | Generic function to accept file path from user using GUI. Returns the filepath value in string format.Default allows all files. |
| cf.gui_get_consent_from_user() | msgForUser="Continue ?" | Generic function to get consent from user using GUI. Returns the string 'yes' or 'no' |
| cf.gui_get_dropdownlist_values_from_user() | msgForUser=" ", dropdown_list=[], multi_select=True | Generic function to accept one of the drop-down value from user using GUI. Returns all chosen values in list format. |
| cf.gui_get_excel_sheet_header_from_user() | msgForUser=" " | Generic function to accept excel path, sheet name and header from user using GUI. Returns all these values in dictionary format. |
| cf.gui_get_folder_path_from_user() | msgForUser="the folder : " | Generic function to accept folder path from user using GUI. Returns the folder path value in string format. |

<br>
<img src="https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/Functions%20Light%20GIFs/gui_function.gif?raw=true" height="400">

----

* ### 4 functions on Mouse Operations:

| Function | Accepted Parameters | Description |
| :--------: | :----: | :----------- |
| cf.mouse_click() | x=" ", y=" ", left_or_right="left", no_of_clicks=1  | Clicks at the given X Y Co-ordinates on the screen using ingle / double / triple click(s). Optionally copies selected data to clipboard (works for double / triple clicks) |
| cf.mouse_move() | x=" ", y=" " | Moves the cursor to the given X Y Co-ordinates |
| cf.mouse_drag_from_to() | x1=" ", y1=" ", x2=" ",y2=" ", delay=0.5 | Clicks and drags from X1 Y1 co-ordinates to X2 Y2 Co-ordinates on the screen |
| cf.mouse_search_snip_return_coordinates_x_y() | img=" ", wait=180 | Searches the given image on the screen and returns its center of X Y co-ordinates. |

<br>
<img src="https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/Mouse_Operations.gif?raw=true" height="400">

----

* ### 6 functions on Window Operations (works only in Windows OS):

| Function | Accepted Parameters | Description |
| :--------: | :----: | :----------- |
| cf.window_show_desktop() | None | Minimizes all the applications and shows Desktop. |
| cf.window_get_all_opened_titles_windows() | window_title=" " | Gives the title of all the existing (open) windows. |
| cf.window_activate_and_maximize_windows() | windowName=" " | Activates and maximizes the desired window. |
| cf.window_minimize_windows() | windowName=" " | Activates and minimizes the desired window. | 
| cf.window_close_windows() | windowName=" " | Close the desired window. |
| cf.launch_any_exe_bat_application() | pathOfExeFile=" " | Launches any exe or batch file or excel file etc. |

<br>
<img src="https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/Functions%20Light%20GIFs/Window Operations.gif?raw=true" height="400">

----
    
* ### 5 functions on Window Objects (works only in Windows OS):

| Function | Accepted Parameters | Description |
| :--------: | :----: | :----------- |
| cf.win_obj_open_app() | title, program_path_with_name, file_path_with_name=" ", backend="uia" | Open any windows application. |
| cf.win_obj_get_all_objects() | main_dlg, save=False, file_name_with_path=" " | Print or Save all the windows object elements of an application. |
| cf.win_obj_mouse_click() | main_dlg,title=" ",  auto_id=" ", control_type=" " | Simulate high level mouse clicks on windows object elements. |
| cf.win_obj_key_press() | main_dlg,write, title=" ", auto_id=" ", control_type=" " | Simulate high level key-press on windows object elements. |
| cf.win_obj_get_text() | main_dlg, title=" ",  auto_id=" ", control_type=" ", value = False | Read text from windows object element. |

<br>
<img src="https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/Windows_Object_Operation.gif?raw=true" height="400">

----

* ### 8 functions on Folder Operations:

| Function | Accepted Parameters | Description |
| :--------: | :----: | :----------- |
| cf.folder_read_text_file() | txt_file_path=" " | Reads from a given text file and returns entire contents as a single list |
| cf.folder_write_text_file() | txt_file_path=" ", contents=" " |  Writes given contents to a text file |
| cf.folder_create() | strFolderPath=" " | When you are making leaf directory, if any intermediate-level directory is missing, folder_create() method creates them. |
| cf.folder_create_text_file() | textFolderPath=" ", txtFileName=" " | Creates text file in the given path. |
| cf.folder_get_all_filenames_as_list() | strFolderPath=" ", extension='all' | Get all the files of the given folder in a list. |

Release_Notes.txt
40) ClointFusion version 1.0.8 released on 01-10-2021 11:30 AM
    1) CLI to send whatsapp messages in bulk is now available
    2) Improved ClointFusion notifier
    3) Bug Fixes and Enhancements


39) ClointFusion version 1.0.7 released on 01-10-2021 11:30 AM
    1) ClointFusion notifier added
    2) Bug Fixes and Enhancements

38) ClointFusion version 1.0.0 - 1.0.6 released on 16-10-2021 11:30 AM
    1) Bug Fixes and Enhancements
    2) Self-test is now guided by Bol

37) ClointFusion version 0.1.45 released on 01-10-2021 11:30 AM
    1) RST file updated for https://readthedocs.org/projects/clointfusion
    2) BOL UI Updated - Murali
    3) DOST Enhancements

36) ClointFusion version 0.1.40 to 0.1.44 released on 23-09-2021 11:30 PM
    1) DOST Working in Browser using Client
    2) Bol got new features
    3) Bug Fixes and Enhancements

35) ClointFusion version 0.1.39 to 0.1.40 released on 18-09-2021 11:30 AM
    1) Bug Fixes and Enhancements

34) ClointFusion version 0.1.38 released on 16-09-2021 18:30 PM
    1) Added new feature Bol (Talk) 
    2) Bug Fixes and Enhancements

33) ClointFusion version 0.1.37 released on 12-09-2021 17:00 PM
    1) Basic Version of DOST Helper / Client Added as a Tray Menu

32) ClointFusion version 0.1.35 - 0.1.36 released on 08-09-2021 20:30 PM
    1) PSTRAY added
    2) WORK HOUR Monitor Feature Added
    3) Bug Fixes and Enhancements Done

31) ClointFusion version 0.1.32 - 0.1.34 released on 03-09-2021 20:30 PM
    1) Basic Version of Work Hour Monitor added
    2) Few Bug Fixes and Enhancements

30) ClointFusion version 0.1.29 - 0.1.31 released on 26-08-2021 20:00 PM
    1) Minor Enhancements
    2) Bug Fixes
    3) Unnecessary imports removed from requirements.txt

29) ClointFusion version 0.1.28 released on 19-08-2021 06:00 PM
    1) CLIs enhanced
    2) New CLIs added
    2) Welcome Message is personalized

28) ClointFusion version 0.1.25 to 0.1.27 released on 15-08-2021 08:00 AM
    1) ChromeDriver issue fixed it better way
    2) Command Line Interface (CLI) linked to Colab Launcher

27) ClointFusion version 0.1.21 to 0.1.24 released on 12-08-2021 05:30 PM
    1) Bug Fixes
    2) Enhancements to few functions

26) ClointFusion version 0.1.20 released on 30-07-2021 09:30 AM
    1) Bug Fixes
    2) Enhancements to few functions

25) ClointFusion version 0.1.17 released on 22-07-2021 11:40 PM
    Bug Fixes

24) ClointFusion version 0.1.16 released on 21-07-2021 11:40 AM
    Code revamped (Contribution by Research Intern Murali Manohar Varma) to make it Compatible with Windows 10, 11, Ubuntu and macOS

23) ClointFusion version 0.1.13 to 0.1.15 released on 09-07-2021 08:30 AM
    1) GIFs added to ReadMe.MD (Contribution by interns@ClointFusion)
    2) Minor bug fixes in few functions
    3) clear_screen, browser_key_press_h, browser_attach_file_h, time_sleep functions contributed

22) ClointFusion version 0.1.12 released on 25-06-2021 11:00 AM
    1) PandasGUI replaced with tabloo PR #21
    2) Code CleanUp

21) ClointFusion version 0.1.10 released on 24-06-2021 17:45 PM
    1) Added new function win10toast
    2) User's can now skip self-test
    3) Removed excel_clean_data & word_cloud_from_url as these are causing some installation challenges in few machines
    4) Helium's Google ChromeDriver issue is now handled in better way

20) ClointFusion version 0.1.8 released on 11-06-2021 17:10 PM
    1) Updated version number in messages
    2) Updated Icon /Logo Files
    3) Fixed minor bug in self-test

19) ClointFusion version 0.1.7 released on 10-06-2021 17:30 PM
    1) excel_split_based_on_row_conditions_unique renamed to excel_split_on_user_defined_conditions
    2) index.rst added ClointFusion master

18) ClointFusion version 0.1.6 released on 10-06-2021 17:00 PM
    1) 11 new functions contributed by interns !
        a) browser_get_title_h
        b) browser_mouse_hover_h
        c)browser_get_dropdown_options_h
        d) browser_select_dropdown_option_h
        e) file_rename
        f) file_get_json_details
        g) excel_change_corrupt_xls_to_xlsx
        h) excel_convert_xls_to_xlsx
        i) excel_apply_template_format_save_to_new
        j) excel_apply_format_as_table
        k) excel_split_on_user_defined_conditions
    2) Existing excel functions now make use of Pandas GUI to show Dataframe in GUI mode
    3) ReadMe file updated

17) ClointFusion version 0.1.5 released on 24-04-2021 10:20 AM
    1) Added new function find() to find and inspect ClointFusion's functions
    2) ReadMe MD, RST file updated
    3) Helium function now connects to user's browser session
    
16) ClointFusion version 0.1.4 released on 23-04-2021 16:00 PM
    1) comtypes package removed from requirements.txt
    2) ReadMe MD & RST file updated
    3) Added ChromeDriver Binary package
    4) time, shutil, sys, datetime, subprocess, traceback are made public

15) ClointFusion version 0.1.3 released on 22-04-2021 8:35 AM
    1) Minor Enhancements
    2) Fixed a bug wrt excel_sort
    3) ReadMe file updated
    4) Compatibility issues wrt Ubuntu

14) ClointFusion version 0.1.2 released on 13-04-2021 10:00 PM
    1) Removed unnecessary Packages
    2) Fixed file not a zipped file error
    3) Minor Enhancements

13) ClointFusion version 0.1.1 released on 09-04-2021 10:50 AM
    1) Enhancements wrt Workspace folder selection & Semi-Automatic Mode
    
12) ClointFusion version 0.1.0 released on 06-04-2021 19:51 PM
    1) Gmail Sign added to Self-Test GUI
    2) User gets a report of self Test
    3) Minor Enhancements related to Performance

11) ClointFusion version 0.0.99 released on 26-03-2021 11:08 AM
    1) Minor Enhancements wrt Package Versions

10) ClointFusion version 0.0.98 released on 17-03-2021 05:35 PM
    1) Minor Enhancements wrt Package Versions

9) ClointFusion version 0.0.97 released on 12-03-2021 07:32 AM
    1) Updated ReadMe files
    2) Updated __init__.py
    3) Updated Documentation
    

ClointFusion/Bol.py
import os
import random
import webbrowser
import requests
import datetime
import sys
import time
import wikipedia
import pyttsx3
import speech_recognition as sr
import pywhatkit as kit
from . import ClointFusion as cf
from pathlib import Path
import subprocess
from PIL import Image
from rich.text import Text
from rich import print
from rich.console import Console

#Install pyaudio
try:
    import pyaudio
except:
    sys_version = str(sys.version[0:6]).strip()
    
    if "3.7" in sys_version :
        cmd = "https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/Wheels/PyAudio-0.2.11-cp37-cp37m-win_amd64.whl?raw=true"
    elif "3.8" in sys_version :
        cmd = "https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/Wheels/PyAudio-0.2.11-cp38-cp38-win_amd64.whl?raw=true"
    elif "3.9" in sys_version :
        cmd = "https://github.com/ClointFusion/Image_ICONS_GIFs/blob/main/Wheels/PyAudio-0.2.11-cp39-cp39-win_amd64.whl?raw=true"

    time.sleep(5)

    try:
        os.system("pip install " + cmd)
    except:
        print("Please install appropriate driver from : https://github.com/ClointFusion/Image_ICONS_GIFs/tree/main/Wheels")

    import pyaudio

engine = pyttsx3.init('sapi5')
voices = engine.getProperty('voices')
voice_male_female = random.randint(0,1) # Randomly decide male/female voice
engine.setProperty('voice', voices[voice_male_female].id)
r = sr.Recognizer()
energy_threshold = [3000]
console = Console()

queries = ["current time,","global news,","send whatsapp,","open , minimize , close any application,","Open Gmail,", "play youtube video,","search in google,",'launch zoom meeting,','switch window,','locate on screen,','take selfie,','OCR now,']
latest_queries = ['launch zoom meeting,','switch window,','locate on screen,','take selfie,','OCR now,']

if cf.os_name == "windows":
    clointfusion_directory = r"C:\Users\{}\ClointFusion".format(str(os.getlogin()))
elif cf.os_name == "linux":
    clointfusion_directory = r"/home/{}/ClointFusion".format(str(os.getlogin()))
elif cf.os_name == "darwin":
    clointfusion_directory = r"/Users/{}/ClointFusion".format(str(os.getlogin()))

def text_to_speech(audio):
    if type(audio) is list:
        print(' '.join(audio))
    else:
        print(str(audio))

    engine.say(audio)   
    engine.runAndWait()

def error_try_later():
    text_to_speech("Sorry, i am experiencing some issues, please try later...")

def shuffle_return_one_option(lst_options=[]):
    random.shuffle(lst_options)
    return str(lst_options[0])

def speech_to_text():
    # cf.message_pop_up("listening",1)
    while True:
        with sr.Microphone() as source:
            r.dynamic_energy_threshold = True
            if r.energy_threshold in energy_threshold or r.energy_threshold <= sorted(energy_threshold)[-1]:
                r.energy_threshold = sorted(energy_threshold)[-1]
            else:
                energy_threshold.append(r.energy_threshold)

            r.pause_threshold = 0.6

            r.adjust_for_ambient_noise(source)

            audio=r.listen(source)
            try:
                query = r.recognize_google(audio)
                print(f"You Said : {query}")
                
                cf.clear_screen()
                print("ClointFusion Bol is here to help you")
                return query
                break
            except sr.UnknownValueError:
                pass
            except sr.RequestError as e:
                print("Try Again")

def _play_sound(music_file_path=""):
    try:
        import wave  
        #define stream chunk   
        chunk = 1024  

        #open a wav format music  
        f = wave.open(music_file_path,"rb")  
        #instantiate PyAudio  
        p = pyaudio.PyAudio()  
        #open stream  
        stream = p.open(format = p.get_format_from_width(f.getsampwidth()),  
                        channels = f.getnchannels(),  
                        rate = f.getframerate(),  
                        output = True)  
        
        data = f.readframes(chunk)  

        #play stream  
        while data:  
            stream.write(data)  
            data = f.readframes(chunk)  

        #stop stream  
        stream.stop_stream()  
        stream.close()  

        #close PyAudio  
        p.terminate()
    except Exception as ex:
        print("Unable to play sound" + str(ex))

def play_on_youtube():
    text_to_speech("OK...")
    text_to_speech("Which video ?")
    video_name = speech_to_text().lower() ## takes user speech_to_text 
    text_to_speech("Opening YouTube now, please wait a moment...")
    kit.playonyt(video_name)

def send_WA_MSG():
    text_to_speech("OK...")
    text_to_speech("Whats the message ?")
    msg = speech_to_text().lower() ## takes user speech_to_text 
    if msg not in ["exit", "cancel", "stop"]:
        text_to_speech("Got it, whom to send, please say mobile number without country code")
    else:
        text_to_speech("Sending message is cancelled...")
        return
    num = speech_to_text().lower() ## takes user speech_to_text
    if num not in ["exit", "cancel", "stop"]:
        text_to_speech("Sending message now, please wait a moment")
        
        kit.sendwhatmsg_instantly(phone_no=f"+91{num}",message=str(msg),wait_time=25, tab_close=True, close_time=5)
    else:
        text_to_speech("Sending message is cancelled...")
        return

def google_search():
    text_to_speech("OK...")
    text_to_speech("What to search ?")
    msg = speech_to_text().lower() ## takes user speech_to_text 
    text_to_speech("Searching in Gooogle now, please wait a moment...")

    kit.search(msg)

def welcome(nth):
    hour = datetime.datetime.now().hour
    greeting = " Good Morning ! " if 5<=hour<12 else " Good Afternoon !! " if hour<18 else " Good Evening !"
    choices = ["Hey ! ", "Hi ! ", "Hello ! ", "Dear ! "]
    greeting = random.choice(choices) + str(cf.user_name) + ' !!' + greeting
    text_to_speech(greeting + " How can i assist you ?!")
    
    if nth == 1:
        suggest(options(3, 2))
    elif nth == 2:
        suggest(options(5, 5))
    elif nth == 3:
        suggest(options(5, 1))
    else:
        pass
    
def suggest(suggestions):
    queries = suggestions
    text_to_speech("Try saying...")
    random.shuffle(queries)
    
    text_to_speech(queries)
    quit_options=['bye','quit','exit']
    random.shuffle(quit_options)
    text_to_speech('To quit, just say {}'.format(quit_options[0]))

def help():
    text_to_speech("Here are all the commands i support currently.")
    print("All commands")
    print(queries)
    text_to_speech("Try some latest commands.")
    print("Latest commands")
    print(latest_queries)
    print("\n")

def options(total=5, latest=3):
    remaining = [q for q in queries if q not in latest_queries]
    custom_list = []
    latest_done = False
    try :
      for i in range(latest+1):
        custom_list.append(latest_queries[i])
      latest_done = True
    except IndexError:
      latest_done = True
    if latest_done:
      for i in range(total - len(custom_list)):
        custom_list.append(remaining[i])
    random.shuffle(custom_list)
    return custom_list

def trndnews(): 
    url = "http://newsapi.org/v2/top-headlines?country=in&apiKey=59ff055b7c754a10a1f8afb4583ef1ab"
    page = requests.get(url).json()
    article = page["articles"]
    results = [ar["title"] for ar in article]
    for i in range(len(results)): 
        print(i + 1, results[i])
    text_to_speech("Here are the top trending news....!!")
    text_to_speech("Do yo want me to read!!!")
    reply = speech_to_text().lower()
    reply = str(reply)
    if reply == "yes":
        text_to_speech(results)
    else:
        text_to_speech('ok!!!!')

def capture_photo(ocr=False):
    try:
        subprocess.run('start microsoft.windows.camera:', shell=True)

        if ocr:
            time.sleep(4)
        else:
            time.sleep(1)

        img=cf.pg.screenshot()
        time.sleep(1)
        
        img.save(Path(os.path.join(clointfusion_directory, "Images","Selfie.PNG")))                    
        subprocess.run('Taskkill /IM WindowsCamera.exe /F', shell=True)
    except Exception as ex:
        print("Error in capture_photo " + str(ex))

def bol_main():
    query_num = 5
    with console.status("Listening...\n") as status:
        while True:
            query = speech_to_text().lower() ## takes user speech_to_text
            
            if 'name' in query:
                text_to_speech("I am ClointFusion's BOL....")

            ### time
            elif 'time' in query:
                time = datetime.datetime.now().strftime('%I:%M %p')
                text_to_speech("It's " + str(time))
            
            elif any(x in query for x in ["help","commands", "list of commands", "what can you do",]):
                help()
            
            ### celebrity
            elif 'who is' in query:
                try:
                    query = query.replace('who is',"")
                    text_to_speech(wikipedia.summary(query,2))
                except:
                    text_to_speech("Please use a complete word...")

            #Send WA MSG
            elif any(x in query for x in ["send whatsapp","whatsapp","whatsapp message"]): 
                try:
                    send_WA_MSG()
                except:
                    error_try_later()

            #Play YouTube Video
            elif any(x in query for x in ["youtube","play video","video song","youtube video"]): 
                try:
                    play_on_youtube()
                except:
                    error_try_later()

            #Search in Google
            elif any(x in query for x in ["google search","search in google"]): 
                try:
                    google_search()
                except:
                    error_try_later()

            #Open gmail
            elif any(x in query for x in ["gmail","email"]): 
                try:
                    webbrowser.open_new_tab("http://mail.google.com")
                except:
                    error_try_later()

            #open camera
            elif any(x in query for x in ["launch camera","open camera"]): 
                try:
                    subprocess.run('start microsoft.windows.camera:', shell=True)
                except:
                    os.startfile('microsoft.windows.camera:')

            ### close camera
            elif any(x in query for x in ["close camera"]): 
                subprocess.run('Taskkill /IM WindowsCamera.exe /F', shell=True)

            ### news
            elif 'news' in query:
                trndnews() 

            #Clap
            elif any(x in query for x in ["clap","applause","shout","whistle"]):
                _play_sound((str(Path(os.path.join(clointfusion_directory,"Logo_Icons","Applause.wav")))))

            elif any(x in query for x in ["bye","quit","stop","exit"]):
                exit_say_choices=["Have a good day! ","Have an awesome day!","I hope your day is great!","Today will be the best!","Have a splendid day!","Have a nice day!","Have a pleasant day!"]
                text_to_speech(shuffle_return_one_option(exit_say_choices))
                break

            elif "dost" in query:
                try:
                    cf.browser_activate('http://dost.clointfusion.com')
                except:
                    pass

            elif any(x in query for x in ["open notepad","launch notepad"]):
                try:
                    cf.launch_any_exe_bat_application("notepad")
                except:
                    pass

            elif any(x in query for x in ["open application","launch application","launch app","open app"]):
                try:
                    text_to_speech('OK, which application to open?')
                    app_name = speech_to_text().lower() ## takes user speech_to_text 
                    cf.launch_any_exe_bat_application(app_name)
                except:
                    pass

            #Switch to window
            elif any(x in query for x in ["switch window","toggle window","activate window","maximize window"]): 
                try:
                    text_to_speech('OK, whats the window name?')
                    windw_name = speech_to_text().lower() ## takes user speech_to_text 
                    cf.window_activate_and_maximize_windows(windw_name)
                except:
                    error_try_later()

            #Search in window / browser
            elif any(x in query for x in ["find on screen","search on screen", "locate on screen"]): 
                try:
                    text_to_speech('OK, what to find ?')
                    query = speech_to_text().lower() ## takes user speech_to_text 
                    cf.find_text_on_screen(searchText=query,delay=0.1, occurance=1,isSearchToBeCleared=False)
                except:
                    error_try_later()

            elif any(x in query for x in ["minimize all","minimize window","show desktop"]):
                try:
                    cf.window_show_desktop()
                except:
                    pass              

            elif any(x in query for x in ["minimize window","minimize application"]):
                try:
                    text_to_speech('OK, which window to minimize?')
                    windw_name = speech_to_text().lower() ## takes user speech_to_text 
                    cf.window_minimize_windows(windw_name)
                except:
                    pass 

            elif any(x in query for x in ["close application","close window"]):
                try:
                    text_to_speech('OK, which application to close?')
                    app_name = speech_to_text().lower() ## takes user speech_to_text 
                    cf.window_close_windows(app_name)
                except:
                    pass

            elif any(x in query for x in ["launch meeting","zoom meeting"]):
                try:
                    webbrowser.open_new_tab("https://us02web.zoom.us/j/85905538540?pwd=b0ZaV3c2bC9zK3I1QXNjYjJ3Q0tGdz09")
                except:
                    pass

            elif "close google chrome" in query:
                try:
                    cf.browser_quit_h()
                except:
                    pass

            elif any(x in query for x in ["take pic","take selfie","take a pic","take a selfie"]):
                try:
                    smile_say=["OK, Smile Please !","OK, Please look at the Camera !","OK, Say Cheese !","OK, Sit up straight !"]

                    text_to_speech(shuffle_return_one_option(smile_say))

                    capture_photo()
                    text_to_speech("Thanks, I saved your photo. Do you want me to open ?")
                    yes_no = speech_to_text().lower() ## takes user speech_to_text 
                    if yes_no in ["yes", "yah", "ok"]:
                        cf.launch_any_exe_bat_application(Path(os.path.join(clointfusion_directory, "Images","Selfie.PNG")))
                    else:
                        pass
                except:
                    error_try_later()

            elif any(x in query for x in ["clear screen","clear","clear terminal","clean", "clean terminal","clean screen",]):
                try:
                    cf.clear_screen()
                    print("ClointFusion Bol is here to help.")
                except Exception as ex:
                    print("Error in clearing " + str(ex))
                    error_try_later()
            
            elif 'ocr' in query:
                try:
                    ocr_say=["OK, Let me scan !","OK, Going to scan now","Please show me the image"]

                    text_to_speech(shuffle_return_one_option(ocr_say))

                    capture_photo(ocr=True)

                    ocr_img_path = Path(os.path.join(clointfusion_directory, "Images","Selfie.PNG"))

                    imageObject = Image.open(ocr_img_path)

                    corrected_image = imageObject.transpose(Image.FLIP_LEFT_RIGHT)

                    corrected_image.save(ocr_img_path)

                    text_to_speech(shuffle_return_one_option(["OK, performing OCR now","Give me a moment","abracadabra","Hang on"]))

                    ocr_result = cf.ocr_now(ocr_img_path)
                    print(ocr_result)
                    text_to_speech(ocr_result)

                except Exception as ex:
                    print("Error in OCR " + str(ex))
                    error_try_later()

            elif any(x in query for x in ["thanks","thank you"]):
                choices = ["You're welcome","You're very welcome.","That's all right.","No problem.","No worries.","Don't mention it.","It's my pleasure.","My pleasure.","Glad to help.","Sure!",""]
                
                text_to_speech(shuffle_return_one_option(choices))            

            elif any(x in query for x in ["shutdown my","turn off","switch off"]):
                try:
                    text_to_speech('Do you want to Shutdown ? Are you sure ?')
                    yes_no = speech_to_text().lower() ## takes user speech_to_text 
                    if yes_no in ["yes", "yah", "ok"]:
                        text_to_speech("OK, Shutting down your machine")

                        os.system('shutdown -s')
                    
                except:
                    error_try_later()

            else:
                query_num += 1
            
                if query_num % 6 == 1:
                    options()

bol_config = clointfusion_directory + "\Config_Files\_bol.txt"

try:
    if os.path.exists(bol_config):
        with open(bol_config, 'r') as fp:
            nth = fp.readline(1)
        with open(bol_config, 'w') as fp:
            fp.write(str(int(nth) + 1))
        welcome(int(nth))
    else:
        with open(bol_config, 'w') as fp:
            fp.write("1")
            welcome(1)
except:
    pass


bol_main()
